services/brig/src/Brig/ZAuth.hs

```diff
@@ -210,21 +296,58 @@ newProviderToken pid = liftZAuth $ do
         let ProviderTokenTimeout ttl = z^.settings.providerTokenTimeout
         in ZC.providerToken ttl (toUUID pid)
 
+mkLegalHoldUserToken :: MonadZAuth m => UserId -> Word32 -> UTCTime -> m LegalHoldUserToken
+mkLegalHoldUserToken u r t = liftZAuth $ do
+    z <- ask
+    liftIO $ ZC.runCreate (z^.private) (z^.settings.keyIndex) $
+       ZC.newToken (utcTimeToPOSIXSeconds t) LU Nothing (mkLegalHoldUser (toUUID u) r)
```

Comment:
The only difference to `mkUserToken'` is the `LU` instead fo the `U`.  I feel there should be a good way to combine the two implementations.  If you:

```haskell
 mkLegalHoldUserToken :: forall (t :: Token) m. MonadZAuth m 
       => UserId -> Word32 -> UTCTime -> m (Token t)
```

And then somehow construct a value from the type-level `LU` or `U`, resp., you have one function that can be called in place of either of these.  Same (probably) for the rest of the class methods.

In fact, here is what I would do:

```haskell
class TokenType {- or some better name -} (a :: Type) where
  tokenTypeVal :: Type  -- i think this will work if you turn on `AllowAmbiguousTypes`; if not use a `Proxy`.

instance TokenType U where
  tokenTypeVal _ = U

instance TokenType LU where
  tokenTypeVal _ = LU
```

and then, instead of as a class method, you can implement `mkUserToken` thusly:

```haskell
mkUserToken :: forall (t :: Type) m. (TokenType t, MonadZAuth m) => UserId -> Word32 -> UTCTime -> m (Token t)
mkUserToken u r t = liftZAuth $ do
    z <- ask
    liftIO $ ZC.runCreate (z^.private) (z^.settings.keyIndex) $
       ZC.newToken (utcTimeToPOSIXSeconds t) (tokenTypeVal @t) Nothing (mkLegalHoldUser (toUUID u) r)
```

Reply:
Thanks! I'll try that.

Reply:
I left a FUTUREWORK for this, as it's not as easy as I had hoped.